File: lab2_task2/main.rb
def count_valid_emails(emails)
    valid_emails = Set.new
    pattern = /^([a-zA-Z0-9\.]{6,30})(\*[a-zA-Z0-9\.]*)?@([a-zA-Z0-9\-\.]+)$/
    
    emails.each do |email|
      match = pattern.match(email)
      unless match
        puts "Неверная почта: #{email}"
        next
      end
      
      username = match[1]
      domain = match[3]
      
      if username.start_with?('.') || username.end_with?('.') || username.include?('..')
        puts "Лишние точки в имени: #{email}"
        next
      end
      
      if domain.start_with?('.') || domain.end_with?('.') || domain.include?('..')
        puts "Лишние точки в домене: #{email}"
        next
      end
      
      clean_username = username.gsub('.', '')
      if clean_username.include?('*')
        clean_username = clean_username.split('*')[0]
        if clean_username.length < 6
          puts "Имя пользователя слишком короткое: #{email}"
          next
        end
      end
      
      valid_emails.add("#{clean_username}@#{domain}")
    end
    
    valid_emails.size
  end
  
  print "Введите адреса почт через пробел: "
  emails = gets.chomp.split
  puts "Количество уникальных адресов: #{count_valid_emails(emails)}"

File: lab2_task2/main.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"regexp"
	"strings"
)

func countValidEmails(emails []string) int {
	validEmails := make(map[string]bool)
	pattern := regexp.MustCompile(`^([a-zA-Z0-9\.]{6,30})(\*[a-zA-Z0-9\.]*)?@([a-zA-Z0-9\-\.]+)$`)
	
	for _, email := range emails {
		matches := pattern.FindStringSubmatch(email)
		if matches == nil {
			fmt.Printf("Неверная почта: %s\n", email)
			continue
		}
		
		username := matches[1]
		domain := matches[3]
		
		if strings.HasPrefix(username, ".") || strings.HasSuffix(username, ".") || strings.Contains(username, "..") {
			fmt.Printf("Лишние точки в имени: %s\n", email)
			continue
		}
		
		if strings.HasPrefix(domain, ".") || strings.HasSuffix(domain, ".") || strings.Contains(domain, "..") {
			fmt.Printf("Лишние точки в домене: %s\n", email)
			continue
		}
		
		cleanUsername := strings.ReplaceAll(username, ".", "")
		if strings.Contains(cleanUsername, "*") {
			parts := strings.Split(cleanUsername, "*")
			cleanUsername = parts[0]
			if len(cleanUsername) < 6 {
				fmt.Printf("Имя пользователя слишком короткое: %s\n", email)
				continue
			}
		}
		
		validEmails[cleanUsername+"@"+domain] = true
	}
	
	return len(validEmails)
}

func main() {
	fmt.Print("Введите адреса почт через пробел: ")
	scanner := bufio.NewScanner(os.Stdin)
	scanner.Scan()
	input := scanner.Text()
	emails := strings.Fields(input)
	fmt.Printf("Количество уникальных адресов: %d\n", countValidEmails(emails))
}